lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
+                const filename = ConfigFile.getFilenameForDirectory(PERSONAL_CONFIG_DIR);
 
-        if (config) {
-            debug("Using config from cache");
-            return config;
+                if (filename) {
+                    debug("Using personal config");
+                    config = ConfigFile.loadCached(filename);
+                }
+            }
+            this.personalConfig = config || null;
         }
 
-        // Step 1: Determine user-specified config from .eslintrc.* and package.json files
+        return this.personalConfig;
+    }
+
+    /**
+     * Builds a hierarchy of config objects, including the base config, all local configs from the directory tree,
+     * and a config file specified on the command line, if applicable.
+     * @param {string} directory a file in whose directory we start looking for a local config
+     * @returns {Object[]} The config objects
+     * @private
+     */
+    getConfigHierarchy(directory) {
+        let configs;
+
+        debug(`Constructing config file hierarchy for ${directory}`);
+
+        // Step 1: Always include baseConfig
+        configs = [this.baseConfig];
+
+        // Step 2: Add user-specified config from .eslintrc.* and package.json files
         if (this.useEslintrc) {
             debug("Using .eslintrc and package.json files");
-            userConfig = getLocalConfig(this, directory);
+            configs = configs.concat(this.getLocalConfigHierarchy(directory));
         } else {
             debug("Not using .eslintrc or package.json files");
-            userConfig = {};
         }
 
-        // Step 2: Create a copy of the baseConfig
-        config = ConfigOps.merge({}, this.baseConfig);
+        // Step 3: Merge in command line config file
+        if (this.useSpecificConfig) {
+            debug("Using command line config file");
+            configs.push(this.useSpecificConfig);
+        }
 
-        // Step 3: Merge in the user-specified configuration from .eslintrc and package.json
-        config = ConfigOps.merge(config, userConfig);
+        return configs;
+    }
 
-        // Step 4: Merge in command line config file
-        if (this.useSpecificConfig) {
-            debug("Merging command line config file");
+    /**
+     * Gets a list of config objects extracted from local config files that apply to the current directory, in
+     * descending order, beginning with the config that is highest in the directory tree.
+     * @param {string} directory The directory to start looking in for local config files.
+     * @returns {Object[]} The shallow local config objects, or an empty array if there are no local configs.
+     * @private
+     */
+    getLocalConfigHierarchy(directory) {
+        const localConfigFiles = this.findLocalConfigFiles(directory),
+            projectConfigPath = ConfigFile.getFilenameForDirectory(this.options.cwd),
+            searched = [],
+            configs = [];
+        let rootPath,
+            cache;
```

Comment:
Nitpick: Can you give this variable a more descriptive name? What is it caching?